File: spark_scripts/moviesrating.py
from pyspark import SparkContext, SparkConf
import sys
import time

start_time = time.time()


conf = SparkConf().setAppName("moviesrating")
sc = SparkContext(conf = conf)

URI = sc._gateway.jvm.java.net.URI
Path = sc._gateway.jvm.org.apache.hadoop.fs.Path
FileSystem = sc._gateway.jvm.org.apache.hadoop.fs.FileSystem
fs = FileSystem.get(URI("hdfs://localhost:9000"), sc._jsc.hadoopConfiguration())


ratingRDD = sc.textFile("hdfs://localhost:9000/cloud/input/rating.csv")
ratingRDDMap = ratingRDD.map(lambda rec: (rec.split(",")[1], rec))

moviesRDD = sc.textFile("hdfs://localhost:9000/cloud/input/movies.csv")
moviesRDDMap = moviesRDD.map(lambda rec: (rec.split(",")[0], rec))

ratingmoviesRDDMap=ratingRDDMap.join(moviesRDDMap)
ratingmoviesRDDMapF = ratingmoviesRDDMap

mapRating = {
  "userid": 0,
  "movieid": 1,
  "rating": 2,
  "timestamp": 3,
}

mapMovies =	{
  "movieid":0,
  "title":1,
  "releasedate":2,
  "unknown": 3,
  "Action": 4,
  "Adventure": 5,
  "Animation": 6,
  "Children": 7,
  "Comedy": 8,
  "Crime": 9,
  "Documentary": 10,
  "Drama": 11,
  "Fantasy": 12,
  "Film_Noir": 13,
  "Horror": 14,
  "Musical": 15,
  "Mystery": 16,
  "Romance": 17,
  "Sci_Fi": 18,
  "Thriller": 19,
  "War": 20,
  "Western": 21,

}

def main():
  
  arguments = sys.argv[1:]
  count = len(arguments)
  if count == 2:
    attribute = arguments[0]
    val = arguments[1]
    attributeIndexRating = mapRating.get(attribute)
    attributeIndexMovies = mapMovies.get(attribute)
    if attributeIndexRating != None:
      ratingmoviesRDDMapF = ratingmoviesRDDMap.filter(lambda row: int(row[1][0].split(",")[attributeIndexRating]) == int(val))
    elif attributeIndexMovies != None:
      if attribute == 'title' or attribute == 'releasedate':
        ratingmoviesRDDMapF = ratingmoviesRDDMap.filter(lambda row: (val in row[1][1].split(",")[attributeIndexMovies]))
      else:
        ratingmoviesRDDMapF = ratingmoviesRDDMap.filter(lambda row: int(row[1][1].split(",")[attributeIndexMovies]) == int(val))
    else:
      ratingmoviesRDDMapF = ratingmoviesRDDMap
  else:
    ratingmoviesRDDMapF = ratingmoviesRDDMap
  
  #for i in ratingmoviesRDDMapF.collect():
  #  print(i)
  
  fs.delete(Path('/cloud/output/spark/moviesrating') ,True)
  
  ratingmoviesRDDMapF.coalesce(1, shuffle = True).saveAsTextFile("hdfs://localhost:9000/cloud/output/spark/moviesrating")
  
  execution_time = time.time() - start_time
  f = open("extime.txt", "w")
  f.write(str(execution_time))
  f.close()

if __name__== "__main__":
  main()

File: Hadoop_Map_reduce_code/join/src/main/java/hadoop/join/MoviesRating.java
package hadoop.join;

import java.net.URI;
import java.util.HashMap;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.lib.input.MultipleInputs;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

public class MoviesRating {

	
	public static void main(String[] args) throws Exception{

		Configuration conf = new Configuration();
		
		HashMap<String, Integer> mapRating = new HashMap<>(); 
		mapRating.put("userid",0);
		mapRating.put("movieid",1);
		mapRating.put("rating",2);
		mapRating.put("timestamp",3);
		
		HashMap<String, Integer> mapMovies = new HashMap<>(); 
		mapMovies.put("movieid",0);
		mapMovies.put("title",1);
		mapMovies.put("releasedate",2);
		mapMovies.put("unknown",3);
		mapMovies.put("Action",4);
		mapMovies.put("Adventure",5);
		mapMovies.put("Animation",6);
		mapMovies.put("Children",7);
		mapMovies.put("Comedy",8);
		mapMovies.put("Crime",9);
		mapMovies.put("Documentary",10);
		mapMovies.put("Drama",11);
		mapMovies.put("Fantasy",12);
		mapMovies.put("Film_Noir",13);
		mapMovies.put("Horror",14);
		mapMovies.put("Musical",15);
		mapMovies.put("Mystery",16);
		mapMovies.put("Romance",17);
		mapMovies.put("Sci_Fi",18);
		mapMovies.put("Thriller",19);
		mapMovies.put("War",20);
		mapMovies.put("Western",21);
		
		
		
		String attribute="DEFAULT";
		String val = "DEFAULT";
		if(args.length==2)
		{
			attribute=args[0];
			val=args[1];
		}

		if(mapRating.containsKey(attribute))
		{
			conf.set("table1","rating");
			conf.set("attributeIndex1", String.valueOf(mapRating.get(attribute)));
			conf.set("val1", val);
		}
		
		if(mapMovies.containsKey(attribute))
		{
			conf.set("table2","movies");
			conf.set("attributeIndex2", String.valueOf(mapMovies.get(attribute)));
			conf.set("val2", val);
		}
		
		Job job = new Job(conf, "Rating Movies");
		 job.setJarByClass(MoviesRating.class);
		 job.setReducerClass(RatingMoviesReducer.class);
		 job.setOutputKeyClass(Text.class);
		 job.setOutputValueClass(Text.class);
		 
		 
		 
		 MultipleInputs.addInputPath(job, new Path("/cloud/input/rating.csv"),TextInputFormat.class, RatingMapperMR.class);
		 MultipleInputs.addInputPath(job, new Path("/cloud/input/movies.csv"),TextInputFormat.class, MoviesMapperMR.class);
		 Path outputPath = new Path("/cloud/output/moviesrating");
		  
		 FileOutputFormat.setOutputPath(job, outputPath);
		 
		 FileSystem hdfs = FileSystem.get(URI.create("hdfs://localhost:9000"),conf);
			
			// delete existing directory
			if (hdfs.exists(outputPath)) {
			    hdfs.delete(outputPath, true);
			}
		 
		 
		 
		 
		 System.exit(job.waitForCompletion(true) ? 0 : 1);
		
	}
	
	
}
